R/approx_distance.R

```diff
@@ -0,0 +1,233 @@
+      # indicator whether the entry is from F or G
+      a <- a0[order(q0)]
+      q <- q0[order(q0)]
+      tau <- c(tau_F, tau_G)[order(q0)]
+      diffs_q <- diff(q)
+      # prob level vectors
+      tau_F_v <- a * tau
+      tau_F_v2 <- sapply(1:(N + M),
+                         function(x)
+                           ifelse(
+                             tau_F_v[x] != 0,
+                             tau_F_v[x],
+                             ifelse(min(which(tau_F_v != 0)) > 1 &&
+                                      x <= min(which(tau_F_v != 0)), 0,
+                                    tau_F_v[max(which(tau_F_v[1:x] !=
+                                                        0))])
+                           ))
+      tau_G_v <- abs(a - 1) * tau
+      tau_G_v2 <- sapply(1:(N + M),
+                         function(x)
+                           ifelse(
+                             tau_G_v[x] != 0,
+                             tau_G_v[x],
+                             ifelse(min(which(tau_G_v != 0)) > 1 &&
+                                      x <= min(which(tau_G_v != 0)), 0,
+                                    tau_G_v[max(which(tau_G_v[1:x] !=
+                                                        0))])
+                           ))
+      if (approx_rule == "left-sided") {
+        cvm <-
+          sum(((tau_F_v2[1:(N + M) - 1] - tau_G_v2[1:(N + M) - 1]) ^ 2) * diffs_q)
+      } else if (approx_rule == "trapezoid") {
+        cvm <-
+          sum((((tau_F_v2[1:(N + M) - 1] - tau_G_v2[1:(N + M) - 1]) ^ 2 + (tau_F_v2[2:(N +
+                                                                                         M)] - tau_G_v2[2:(N + M)]) ^ 2
+          ) / 2) * diffs_q)
+      }
+      metric <- cvm
+    }
+    return(metric)
+  }
+
+#' Approximate pairwise Cramér distance for a single target,
+#' date, location data frame
+#'
+#' @param single_frame a data frame containing correctly-formatted forecasts
+#' for a single target, location, target_end_date
+#' @param approx_rule a string specifying which formula to use
+#' for approximation. Valid rules are "left-sided" and "trapezoid".
+#' The default rule is "trapezoid".
+#' @return a data frame with approximated pairwise Cramér distances
+#' for all possible pairs of models
+#'
+#' @export
+#'
+cd_combination <- function(single_frame, end_date, approx_rule) {
+  if (approx_rule == "left-sided" |
+      approx_rule == "trapezoid") {
+    tau_F <- tau_G <- unique(single_frame$quantile)
+  } else {
+    tau_F <- tau_G <- NULL
+  }
+  # remove any models with NA values
+  single_tarloc <-
+    single_frame[, colSums(is.na(single_frame)) == 0] %>%
+    dplyr::filter(target_end_date == end_date) %>%
+    dplyr::arrange(quantile)
+  # pairwise column calculation
+  tmp <- single_tarloc %>%
+    dplyr::select(
+      -c(
+        "target_variable",
+        "target_end_date",
+        "location",
+        "type",
+        "quantile",
+        "horizon",
```

Comment:
Are multiple horizons getting mushed together here? it's not clear to me how this is working.